File: native/android/VueNativeCore/src/main/kotlin/com/vuenative/core/Modules/AnimationModule.kt
package com.vuenative.core

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.Keyframe
import android.animation.ObjectAnimator
import android.animation.PropertyValuesHolder
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import android.view.animation.LinearInterpolator

class AnimationModule : NativeModule {
    override val moduleName = "Animation"
    private var context: Context? = null
    private var bridge: NativeBridge? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun initialize(context: Context, bridge: NativeBridge) {
        this.context = context
        this.bridge = bridge
    }

    override fun invoke(method: String, args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        when (method) {
            "timing" -> handleTiming(args, bridge, callback)
            "spring" -> handleSpring(args, bridge, callback)
            "keyframe" -> handleKeyframe(args, bridge, callback)
            "sequence" -> handleSequence(args, bridge, callback)
            "parallel" -> handleParallel(args, bridge, callback)
            "fadeIn" -> {
                val nodeId = StyleEngine.toInt(args.getOrNull(0), -1)
                val view = bridge.nodeViews[nodeId] ?: run {
                    callback(null, null)
                    return
                }
                mainHandler.post {
                    view.alpha = 0f
                    view.visibility = View.VISIBLE
                    view.animate().alpha(1f).setDuration(300).withEndAction { callback(null, null) }.start()
                }
            }
            "fadeOut" -> {
                val nodeId = StyleEngine.toInt(args.getOrNull(0), -1)
                val view = bridge.nodeViews[nodeId] ?: run {
                    callback(null, null)
                    return
                }
                mainHandler.post {
                    view.animate().alpha(0f).setDuration(300).withEndAction {
                        view.visibility = View.INVISIBLE
                        callback(null, null)
                    }.start()
                }
            }
            else -> callback(null, "Unknown animation method: $method")
        }
    }

    // ── timing(viewId, toStyles, options) ──────────────────────────────────
    // args[0]: nodeId (Int)
    // args[1]: toStyles map { "opacity": 1, "translateX": 100 }
    // args[2]: options { duration, delay, easing }

    private fun handleTiming(args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        val nodeId = StyleEngine.toInt(args.getOrNull(0), -1)
        val toStyles = toStringKeyMap(args.getOrNull(1)) ?: emptyMap()
        val options = toStringKeyMap(args.getOrNull(2)) ?: emptyMap()

        val duration = StyleEngine.toInt(options["duration"], 300).toLong()
        val delay = StyleEngine.toInt(options["delay"], 0).toLong()
        val easing = options["easing"]?.toString() ?: "easeInOut"

        val view = bridge.nodeViews[nodeId] ?: run {
            callback(null, "timing: view $nodeId not found")
            return
        }

        mainHandler.post {
            val animators = buildPropertyAnimators(view, toStyles)
            if (animators.isEmpty()) {
                callback(true, null)
                return@post
            }

            val interpolator = resolveInterpolator(easing)
            val set = AnimatorSet()
            set.playTogether(animators)
            set.duration = duration
            set.startDelay = delay
            set.interpolator = interpolator
            set.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(a: Animator) {
                    callback(true, null)
                }
            })
            set.start()
        }
    }

    // ── spring(viewId, toStyles, options) ──────────────────────────────────
    // args[0]: nodeId
    // args[1]: toStyles map
    // args[2]: options { damping/tension/friction, duration, delay }

    private fun handleSpring(args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        val nodeId = StyleEngine.toInt(args.getOrNull(0), -1)
        val toStyles = toStringKeyMap(args.getOrNull(1)) ?: emptyMap()
        val options = toStringKeyMap(args.getOrNull(2)) ?: emptyMap()

        val duration = StyleEngine.toInt(options["duration"], 500).toLong()
        val delay = StyleEngine.toInt(options["delay"], 0).toLong()

        val view = bridge.nodeViews[nodeId] ?: run {
            callback(null, "spring: view $nodeId not found")
            return
        }

        mainHandler.post {
            val animators = buildPropertyAnimators(view, toStyles)
            if (animators.isEmpty()) {
                callback(true, null)
                return@post
            }

            // Android doesn't have built-in spring for ObjectAnimator pre-API 23 DynamicAnimation.
            // Use overshoot interpolator to approximate spring feel.
            val set = AnimatorSet()
            set.playTogether(animators)
            set.duration = duration
            set.startDelay = delay
            set.interpolator = android.view.animation.OvershootInterpolator(1.5f)
            set.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(a: Animator) {
                    callback(true, null)
                }
            })
            set.start()
        }
    }

    // ── keyframe(viewId, keyframeSteps, options) ───────────────────────────
    // args[0]: nodeId
    // args[1]: List<Map> keyframe steps [{ offset: 0.0, opacity: 1 }, ...]
    // args[2]: options { duration }

    private fun handleKeyframe(args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        val nodeId = StyleEngine.toInt(args.getOrNull(0), -1)
        val stepsRaw = args.getOrNull(1) as? List<*> ?: run {
            callback(null, "keyframe: invalid keyframes")
            return
        }
        val options = toStringKeyMap(args.getOrNull(2)) ?: emptyMap()
        val duration = StyleEngine.toInt(options["duration"], 300).toLong()

        val view = bridge.nodeViews[nodeId] ?: run {
            callback(null, "keyframe: view $nodeId not found")
            return
        }

        // Parse keyframe steps into Map<String, List<Pair<Float, Float>>> (property -> [(offset, value)])
        val propertyKeyframes = mutableMapOf<String, MutableList<Pair<Float, Float>>>()
        for (raw in stepsRaw) {
            val step = toStringKeyMap(raw) ?: continue
            val offset = StyleEngine.toFloat(step["offset"], 0f)
            for ((key, value) in step) {
                if (key == "offset") continue
                val floatVal = StyleEngine.toFloat(value, Float.NaN)
                if (floatVal.isNaN()) continue
                propertyKeyframes.getOrPut(key) { mutableListOf() }.add(offset to floatVal)
            }
        }

        if (propertyKeyframes.isEmpty()) {
            callback(null, null)
            return
        }

        mainHandler.post {
            val animators = mutableListOf<Animator>()

            for ((propKey, frames) in propertyKeyframes) {
                val androidProps = mapPropertyName(propKey)
                for (androidProp in androidProps) {
                    val kfs = frames.map { (offset, value) ->
                        Keyframe.ofFloat(offset, value)
                    }.toTypedArray()
                    val pvh = PropertyValuesHolder.ofKeyframe(androidProp, *kfs)
                    val animator = ObjectAnimator.ofPropertyValuesHolder(view, pvh)
                    animator.duration = duration
                    animators.add(animator)
                }
            }

            if (animators.isEmpty()) {
                callback(null, null)
                return@post
            }

            val set = AnimatorSet()
            set.playTogether(animators)
            set.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(a: Animator) {
                    callback(null, null)
                }
            })
            set.start()
        }
    }

    // ── sequence(animations) ───────────────────────────────────────────────
    // args[0]: List<Map> animations [{ type, viewId, toStyles, options }, ...]

    private fun handleSequence(args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        val animationsRaw = args.getOrNull(0) as? List<*> ?: run {
            callback(null, "sequence: invalid args")
            return
        }
        val animations = animationsRaw.mapNotNull { toStringKeyMap(it) }
        if (animations.isEmpty()) {
            callback(null, null)
            return
        }

        runSequenceStep(animations, 0, bridge, callback)
    }

    private fun runSequenceStep(
        animations: List<Map<String, Any?>>,
        index: Int,
        bridge: NativeBridge,
        callback: (Any?, String?) -> Unit
    ) {
        if (index >= animations.size) {
            callback(null, null)
            return
        }

        val animData = animations[index]
        val method = animData["type"]?.toString() ?: "timing"
        val viewId = StyleEngine.toInt(animData["viewId"], 0)
        val toStyles = toStringKeyMap(animData["toStyles"]) ?: emptyMap()
        val options = toStringKeyMap(animData["options"]) ?: emptyMap()

        val subArgs: List<Any?> = listOf(viewId, toStyles, options)
        invoke(method, subArgs, bridge) { _, error ->
            if (error != null) {
                callback(null, error)
                return@invoke
            }
            runSequenceStep(animations, index + 1, bridge, callback)
        }
    }

    // ── parallel(animations) ───────────────────────────────────────────────
    // args[0]: List<Map> animations [{ type, viewId, toStyles, options }, ...]

    private fun handleParallel(args: List<Any?>, bridge: NativeBridge, callback: (Any?, String?) -> Unit) {
        val animationsRaw = args.getOrNull(0) as? List<*> ?: run {
            callback(null, "parallel: invalid args")
            return
        }
        val animations = animationsRaw.mapNotNull { toStringKeyMap(it) }
        if (animations.isEmpty()) {
            callback(null, null)
            return
        }

        val total = animations.size
        var completed = 0
        val lock = Object()

        for (animData in animations) {
            val method = animData["type"]?.toString() ?: "timing"
            val viewId = StyleEngine.toInt(animData["viewId"], 0)
            val toStyles = toStringKeyMap(animData["toStyles"]) ?: emptyMap()
            val options = toStringKeyMap(animData["options"]) ?: emptyMap()

            val subArgs: List<Any?> = listOf(viewId, toStyles, options)
            invoke(method, subArgs, bridge) { _, _ ->
                val allDone: Boolean
                synchronized(lock) {
                    completed++
                    allDone = completed == total
                }
                if (allDone) callback(null, null)
            }
        }
    }

    // ── Helpers ─────────────────────────────────────────────────────────────

    /**
     * Build a list of ObjectAnimator from a toStyles map like { opacity: 1, translateX: 100 }.
     * Each style key maps to one or more Android View properties.
     */
    private fun buildPropertyAnimators(view: View, toStyles: Map<String, Any?>): List<Animator> {
        val animators = mutableListOf<Animator>()
        for ((key, value) in toStyles) {
            val toValue = StyleEngine.toFloat(value, Float.NaN)
            if (toValue.isNaN()) continue
            val androidProps = mapPropertyName(key)
            for (prop in androidProps) {
                animators.add(ObjectAnimator.ofFloat(view, prop, toValue))
            }
        }
        return animators
    }

    /**
     * Map a JS-side property name to Android View property name(s).
     * "scale" maps to both "scaleX" and "scaleY".
     */
    private fun mapPropertyName(propKey: String): List<String> = when (propKey) {
        "opacity" -> listOf("alpha")
        "translateX" -> listOf("translationX")
        "translateY" -> listOf("translationY")
        "scale" -> listOf("scaleX", "scaleY")
        "scaleX" -> listOf("scaleX")
        "scaleY" -> listOf("scaleY")
        "rotate" -> listOf("rotation")
        else -> emptyList()
    }

    /**
     * Resolve an easing string to an Android Interpolator.
     */
    private fun resolveInterpolator(easing: String) = when (easing) {
        "linear" -> LinearInterpolator()
        "ease-in", "easeIn" -> AccelerateInterpolator()
        "ease-out", "easeOut" -> DecelerateInterpolator()
        else -> AccelerateDecelerateInterpolator() // "easeInOut", "ease", default
    }

    /**
     * Safely cast Any? to Map<String, Any?>, handling Map<*, *> from J2V8/JSON deserialization.
     */
    private fun toStringKeyMap(value: Any?): Map<String, Any?>? {
        val map = value as? Map<*, *> ?: return null
        @Suppress("UNCHECKED_CAST")
        return map as Map<String, Any?>
    }
}

File: native/Sources/VueNativeCore/Modules/AnimationModule.swift
#if canImport(UIKit)
import UIKit

/// Native module for UIView animations.
/// Exposes timing, spring, keyframe, sequence, and parallel animations to JS composables.
final class AnimationModule: NativeModule {

    let moduleName = "Animation"

    func invoke(method: String, args: [Any], callback: @escaping (Any?, String?) -> Void) {
        switch method {
        case "timing":
            handleTiming(args: args, callback: callback)
        case "spring":
            handleSpring(args: args, callback: callback)
        case "keyframe":
            guard let viewId = args[safe: 0] as? Int,
                  let keyframesData = args[safe: 1] as? [[String: Any]],
                  let options = args[safe: 2] as? [String: Any] else {
                callback(nil, "Invalid args"); return
            }
            let duration = options["duration"] as? Double ?? 300
            animateKeyframes(viewId: viewId, keyframes: keyframesData, duration: duration / 1000.0, callback: callback)
        case "sequence":
            guard let animationsData = args[safe: 0] as? [[String: Any]] else {
                callback(nil, "Invalid args"); return
            }
            runSequence(animationsData: animationsData, index: 0, callback: callback)
        case "parallel":
            guard let animationsData = args[safe: 0] as? [[String: Any]] else {
                callback(nil, "Invalid args"); return
            }
            runParallel(animationsData: animationsData, callback: callback)
        default:
            callback(nil, "AnimationModule: unknown method '\(method)'")
        }
    }

    func invokeSync(method: String, args: [Any]) -> Any? {
        return nil
    }

    // MARK: - timing(viewId, styles, options)
    // args[0]: Int (viewId)
    // args[1]: [String: Any] (target styles, only numeric values like opacity, transform offsets)
    // args[2]: [String: Any] (options: duration, delay, easing)

    private func handleTiming(args: [Any], callback: @escaping (Any?, String?) -> Void) {
        guard let viewId = args.first.flatMap({ $0 as? Int ?? ($0 as? Double).map(Int.init) }),
              let styles = args.count > 1 ? args[1] as? [String: Any] : [:] else {
            callback(nil, "timing: invalid arguments")
            return
        }

        let options = args.count > 2 ? (args[2] as? [String: Any] ?? [:]) : [:]
        let duration = (options["duration"] as? Double ?? 300) / 1000.0
        let delay = (options["delay"] as? Double ?? 0) / 1000.0
        let easing = options["easing"] as? String ?? "ease"

        let curve: UIView.AnimationOptions
        switch easing {
        case "linear": curve = .curveLinear
        case "ease-in", "easeIn": curve = .curveEaseIn
        case "ease-out", "easeOut": curve = .curveEaseOut
        default: curve = .curveEaseInOut
        }

        DispatchQueue.main.async {
            MainActor.assumeIsolated {
                guard let view = NativeBridge.shared.view(forId: viewId) else {
                    callback(nil, "timing: view \(viewId) not found")
                    return
                }

                UIView.animate(
                    withDuration: duration,
                    delay: delay,
                    options: [curve],
                    animations: { [styles] in
                        MainActor.assumeIsolated {
                            for (key, value) in styles {
                                StyleEngine.apply(key: key, value: value, to: view)
                            }
                        }
                    },
                    completion: { _ in
                        callback(true, nil)
                    }
                )
            }
        }
    }

    // MARK: - spring(viewId, styles, options)
    // args[0]: Int (viewId)
    // args[1]: [String: Any] (target styles)
    // args[2]: [String: Any] (options: damping, stiffness/duration, mass)

    private func handleSpring(args: [Any], callback: @escaping (Any?, String?) -> Void) {
        guard let viewId = args.first.flatMap({ $0 as? Int ?? ($0 as? Double).map(Int.init) }),
              let styles = args.count > 1 ? args[1] as? [String: Any] : [:] else {
            callback(nil, "spring: invalid arguments")
            return
        }

        let options = args.count > 2 ? (args[2] as? [String: Any] ?? [:]) : [:]
        let damping = CGFloat(options["damping"] as? Double ?? 0.7)
        let duration = (options["duration"] as? Double ?? 500) / 1000.0
        let delay = (options["delay"] as? Double ?? 0) / 1000.0

        DispatchQueue.main.async {
            MainActor.assumeIsolated {
                guard let view = NativeBridge.shared.view(forId: viewId) else {
                    callback(nil, "spring: view \(viewId) not found")
                    return
                }

                UIView.animate(
                    withDuration: duration,
                    delay: delay,
                    usingSpringWithDamping: damping,
                    initialSpringVelocity: 0,
                    options: [],
                    animations: { [styles] in
                        MainActor.assumeIsolated {
                            for (key, value) in styles {
                                StyleEngine.apply(key: key, value: value, to: view)
                            }
                        }
                    },
                    completion: { _ in
                        callback(true, nil)
                    }
                )
            }
        }
    }

    // MARK: - keyframe(viewId, keyframes, options)

    private func animateKeyframes(viewId: Int, keyframes: [[String: Any]], duration: TimeInterval, callback: @escaping (Any?, String?) -> Void) {
        DispatchQueue.main.async {
            guard let view = NativeBridge.shared.view(forId: viewId) else {
                callback(nil, "View not found for id \(viewId)"); return
            }

            // Build CAKeyframeAnimation for each animated property
            // Group them with CAAnimationGroup
            let group = CAAnimationGroup()
            group.duration = duration
            group.fillMode = .forwards
            group.isRemovedOnCompletion = false

            var animations: [CAAnimation] = []

            // Collect all property names from keyframes
            var propertyKeys = Set<String>()
            for kf in keyframes { propertyKeys.formUnion(kf.keys.filter { $0 != "offset" }) }

            for propKey in propertyKeys {
                let keyPath: String
                var values: [Any] = []
                var keyTimes: [NSNumber] = []

                switch propKey {
                case "opacity":
                    keyPath = "opacity"
                    for kf in keyframes {
                        if let v = kf[propKey] as? Double {
                            values.append(v)
                            keyTimes.append(NSNumber(value: kf["offset"] as? Double ?? 0))
                        }
                    }
                case "translateX":
                    keyPath = "transform.translation.x"
                    for kf in keyframes {
                        if let v = kf[propKey] as? Double {
                            values.append(v)
                            keyTimes.append(NSNumber(value: kf["offset"] as? Double ?? 0))
                        }
                    }
                case "translateY":
                    keyPath = "transform.translation.y"
                    for kf in keyframes {
                        if let v = kf[propKey] as? Double {
                            values.append(v)
                            keyTimes.append(NSNumber(value: kf["offset"] as? Double ?? 0))
                        }
                    }
                case "scale", "scaleX":
                    keyPath = "transform.scale.x"
                    for kf in keyframes {
                        if let v = kf[propKey] as? Double {
                            values.append(v)
                            keyTimes.append(NSNumber(value: kf["offset"] as? Double ?? 0))
                        }
                    }
                case "scaleY":
                    keyPath = "transform.scale.y"
                    for kf in keyframes {
                        if let v = kf[propKey] as? Double {
                            values.append(v)
                            keyTimes.append(NSNumber(value: kf["offset"] as? Double ?? 0))
                        }
                    }
                default:
                    continue
                }

                if !values.isEmpty {
                    let anim = CAKeyframeAnimation(keyPath: keyPath)
                    anim.values = values
                    anim.keyTimes = keyTimes
                    anim.duration = duration
                    animations.append(anim)
                }
            }

            if animations.isEmpty { callback(nil, nil); return }
            group.animations = animations

            CATransaction.begin()
            CATransaction.setCompletionBlock { callback(nil, nil) }
            view.layer.add(group, forKey: "keyframeAnimation")
            CATransaction.commit()
        }
    }

    // MARK: - sequence([animData])

    private func runSequence(animationsData: [[String: Any]], index: Int, callback: @escaping (Any?, String?) -> Void) {
        guard index < animationsData.count else { callback(nil, nil); return }
        let animData = animationsData[index]
        let method = animData["type"] as? String ?? "timing"
        let viewId = animData["viewId"] as? Int ?? 0
        let toStyles = animData["toStyles"] as? [String: Any] ?? [:]
        let options = animData["options"] as? [String: Any] ?? [:]

        runSingleAnimation(method: method, viewId: viewId, toStyles: toStyles, options: options) { [weak self] _, error in
            if let error = error { callback(nil, error); return }
            self?.runSequence(animationsData: animationsData, index: index + 1, callback: callback)
        }
    }

    private func runSingleAnimation(method: String, viewId: Int, toStyles: [String: Any], options: [String: Any], callback: @escaping (Any?, String?) -> Void) {
        switch method {
        case "timing":
            let argsArr: [Any] = [viewId, toStyles, options]
            invoke(method: "timing", args: argsArr, callback: callback)
        case "spring":
            let argsArr: [Any] = [viewId, toStyles, options]
            invoke(method: "spring", args: argsArr, callback: callback)
        default:
            callback(nil, "Unknown animation type: \(method)")
        }
    }

    // MARK: - parallel([animData])

    private func runParallel(animationsData: [[String: Any]], callback: @escaping (Any?, String?) -> Void) {
        guard !animationsData.isEmpty else { callback(nil, nil); return }
        let total = animationsData.count
        var completed = 0
        let lock = NSLock()

        for animData in animationsData {
            let method = animData["type"] as? String ?? "timing"
            let viewId = animData["viewId"] as? Int ?? 0
            let toStyles = animData["toStyles"] as? [String: Any] ?? [:]
            let options = animData["options"] as? [String: Any] ?? [:]

            runSingleAnimation(method: method, viewId: viewId, toStyles: toStyles, options: options) { _, _ in
                lock.lock()
                completed += 1
                let allDone = completed == total
                lock.unlock()
                if allDone { callback(nil, nil) }
            }
        }
    }
}

// MARK: - Array safe subscript

private extension Array {
    subscript(safe index: Int) -> Element? {
        guard index >= 0 && index < count else { return nil }
        return self[index]
    }
}
#endif
